configure latexmk as the defualt build command

izahn opened the pull request:
This PR configures automatic compilation of LaTeX documents via `latexmk` by default if `latexmk` is available on the system. It also clarifies the use of the `latex-build-command` variable in response to https://github.com/syl20bnr/spacemacs/issues/2462#issuecomment-126143492. With these changes emacs will be configured to compile latex documents in a single step unless the user opts out, or the `latexmk` executable is unavailable.

The changes proposed here require a recent version of https://github.com/tom-tan/auctex-latexmk -- I'm not sure what the expectation is here. Since the changes I propose here will break without the latest version of `auctex-latexmk` should add an `if` expression to handle older versions, or do we expect users to keep packages up to date?

justbur commented:
My preference would be to use your suggestion of using a boolean `latex-build-use-latexMk`. It's much easier than possibly screwing up the capitalization of "LatexMk". If you're worried about backwards compatibility, maybe you could do something simple at the beginning like 

```elisp
(when (string= latex-build-command "LatexMk") (setq latex-build-use-latexMk t))
```

If you're going to keep `latex-build-command`, I also think it would be better to use the actual name of the variable `TeX-command-default` instead, because it's more transparent. 

Regardless, I like the changes.

izahn commented:
Yes, I agree with @justbur that this PR leaves some things unaddressed. However, it configures things so they should just work by default in most cases, and clarifies the documentation to reduce confusion. It does this things without making it any more difficult to clean things up later. In short I think this is a good step in the right direction, and we should take it and then start working on the next step.

izahn commented:
I rebased and moved the `auctex-latexmk` configuration code inside the `use-package` expression. The functionality is unchanged from my original description.

izahn commented:
I know there is a long PR queue, but I'd really like to see this merged. I know it seems like a small change, but I worked hard to get a patch into upstream so this will work, and I'd like to see spacemacs take advantage of it.

izahn commented:
I know spacemacs has been busy over the last month, so maybe this fell through the cracks. Or maybe there is no interest in merging it. Either way it would be nice to get some feedback about whether this PR has a chance of being accepted, or whether I should give up and close it. Thanks.

syl20bnr commented:
Sorry, I'll take a look at it.

syl20bnr commented on layers/+lang/latex/packages.el:
```diff
@@ -77,10 +77,16 @@
+      (progn
+        (add-hook 'LaTeX-mode-hook
+                  (lambda()
+                    (setq TeX-command-default "LatexMk")))
```
Why do you need to put this in a lambda ? Why not just setting it once ?

izahn replied:
I'm not sure, but I think there was a reason. I'll try just setting it once and see if it breaks anything -- if not I'll make that change and update the PR.

izahn replied:
I tested it out and just setting it once doesn't work. Not sure where exactly it gets overwritten, but it does. Adding it to `LaTeX-mode-hook` is the only way I can see to make it actually work. Of course I'm happy if someone can see a cleaner way.

syl20bnr commented on layers/+lang/latex/packages.el:
```diff
@@ -77,10 +77,16 @@
+        (add-hook 'LaTeX-mode-hook
+                  (lambda()
+                    (setq TeX-command-default "LatexMk")))
+        (auctex-latexmk-setup)
```
Why you put this outside of a hook ? It think it will defeat the lazy-loading.

izahn replied:
I'm not sure what hook you are referring to, or what is outside of it. I was mostly trying to keep the structure that already existed, just adding what needed to be added to make `LatexMk` work by default.

syl20bnr replied:
See at line 83 the setup what is in the LaTeX hook.

izahn replied:
Ah I see. Well, `auctex-latexmk-setup` only needs to be called once, it certainly doesn't need to be in `LaTeX-mode-hook`. Based on my very basic understanding of use-package this should be called in a `:config` macro. Is that a correct way to do it that works with lazy-loading?

izahn replied:
I've addressed the lazy loading issue by wrapping the LatexMk configuration in (eval-after-load "tex-mode". Feels wrong since I guess use-package is designed to avoid needing to do that, but I don't understand use-package well enough. At lease the PR now implements LatexMk support without interfering with lazy loading.

izahn commented:
@syl20bnr Thanks for the feedback I appreciate it! I hope I'll be able to address your questions and get this into spacemacs.

izahn commented:
I've addressed the lazy loading issue by wrapping the LatexMk configuration in `(eval-after-load "tex-mode"`. Feels wrong since I guess use-package is designed to avoid needing to do that, but I don't understand use-package well enough. At lease the PR now implements LatexMk support without interfering with lazy loading.

izahn commented:
Sorry for the delay in getting back to this. This PR now enables `latexmk` by default on systems where the executable can be found, without interfering with lazy loading. I've been using it for the past couple of weeks without any problems. I just cleanup up the documentation, and I think this is finally ready to go.

TheBB commented:
Thank you! (For your patience, as well.) I just cherry-picked this in develop, so you can safely delete your branch.

I changed the code a bit in https://github.com/syl20bnr/spacemacs/commit/6887320420cac8c25d40484cf71d906a3212df26. I set `auctex-latexmk-inherit-TeX-PDF-mode` during init instead of deferred, so it can more easily be overwritten. I also moved the `auctex-latexmk` setup into a use-package hook. Finally I removed the hook setup for `TeX-command-default` as it seemed it didn't matter, for the following reasons:

- The binding at `SPC m b` doesn't use `TeX-command-default` but `latex-build-command`.
- After running `SPC m b`, `TeX-command-default` has anyway changed to LatexMk.

I left the current variable in, it's a string and not a boolean flag, because due to its use in `latex/build` it really has many more possible values, not just `LaTeX` or `LatexMk`. I agree we should make wrapper around it so that it becomes case insensitive, though.
